src/Analyzers/Core/Analyzers/NamespaceMatchFolder/AbstractNamespaceMatchFolderDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,196 @@
+        public const string TargetNamespace = "TargetNamespace";
+
+        private static readonly LocalizableResourceString s_localizableTitle = new(
+          nameof(AnalyzersResources.Namespace_does_not_match_folder_structure), AnalyzersResources.ResourceManager, typeof(AnalyzersResources));
+
+        private static readonly LocalizableResourceString s_localizableInsideMessage = new(
+            nameof(AnalyzersResources.Namespace_0_does_not_match_folder_structure_expected_1), AnalyzersResources.ResourceManager, typeof(AnalyzersResources));
+
+        protected AbstractNamespaceMatchFolderDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.NamespaceSyncAnalyzerDiagnosticId,
+                EnforceOnBuildValues.NamespaceSync,
+                CodeStyleOptions2.PreferNamespaceMatchFolderStructure,
+                s_localizableTitle,
+                s_localizableInsideMessage)
+        {
+        }
+    }
+
+    internal abstract class AbstractNamespaceMatchFolderDiagnosticAnalyzer<TNamespaceSyntax> : AbstractNamespaceMatchFolderDiagnosticAnalyzer
+        where TNamespaceSyntax : SyntaxNode
+    {
+        private static readonly SymbolDisplayFormat s_namespaceDisplayFormat = SymbolDisplayFormat
+            .FullyQualifiedFormat
+            .WithGlobalNamespaceStyle(SymbolDisplayGlobalNamespaceStyle.Omitted);
+
+        /// <summary>
+        /// Gets the language specific syntax facts
+        /// </summary>
+        protected abstract ISyntaxFacts GetSyntaxFacts();
+
+        /// <summary>
+        /// Gets the syntax representing the name of a namespace declaration
+        /// </summary>
+        protected abstract SyntaxNode GetNameSyntax(TNamespaceSyntax namespaceDeclaration);
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected void AnalyzeNamespaceNode(SyntaxNodeAnalysisContext context)
+        {
+            // It's ok to not have a rootnamespace property, but if it's there we want to use it correctly
+            context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(RootNamespaceOption, out var rootNamespace);
+
+            // Project directory is a must to correctly get the relative path and construct a namespace
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(ProjectDirOption, out var projectDir)
+                || string.IsNullOrEmpty(projectDir))
+            {
+                return;
+            }
+
+            if (context.Node is TNamespaceSyntax namespaceDecl)
+            {
+                var currentNamespace = GetNamespaceName(context.SemanticModel, namespaceDecl);
+
+                if (IsFileAndNamespaceMismatch(namespaceDecl, rootNamespace, projectDir, currentNamespace, out var targetNamespace) &&
+                    IsFixSupported(context.SemanticModel, namespaceDecl, context.CancellationToken))
```

Comment:
One more question here. There's scenarios we don't report diagnostics because our backing service doesn't support the auto fix (i.e. nest namespace declaration, partial types, etc), but it's still a mismatch. Do we want to show a diagnostic in those cases?

Reply:
Maybe we should show a diagnostic... not sure how we feel about "magic" where a fix shows up or doesn't

Reply:
At least for VB it's consistent on the whole language. If we have some cases where a fix shows and some where it doesn't, I don't know how a user rationalizes relying on it being there.

Reply:
I don't know if there's a precedent, but if we decide to show an diagnostic in unsupported case, maybe we can call it out in the message?

Reply:
i would not show the diagnostic since we can't fix it and that could leave hte user confused as to what they have to do.

Reply:
🤦‍♂️